src/js/integrations/stalltracking.ts

```diff
@@ -0,0 +1,393 @@
+            timestamp: spanEndTimestamp,
+            stats: this._getCurrentStats(transaction),
+          },
+        });
+      }
+    }
+  }
+
+  /**
+   * Get the current stats for a transaction at a given time.
+   */
+  private _getCurrentStats(transaction: Transaction): StallMeasurements {
+    return {
+      stall_count: { value: this._stallCount },
+      stall_total_time: { value: this._totalStallTime },
+      stall_longest_time: {
+        value: this._statsByTransaction.get(transaction)?.longestStallTime ?? 0,
+      },
+    };
+  }
+
+  /**
+   * Start tracking stalls
+   */
+  private _startTracking(): void {
+    if (!this.isTracking) {
+      this.isTracking = true;
+      this._lastIntervalMs = Math.floor(timestampInSeconds() * 1000);
+
+      this._iteration();
+    }
+  }
+
+  /**
+   * Stops the stall tracking interval and calls reset().
+   */
+  private _stopTracking(): void {
+    this.isTracking = false;
+
+    if (this._timeout !== null) {
+      clearTimeout(this._timeout);
+      this._timeout = null;
+    }
+
+    this._reset();
+  }
+
+  /**
+   * Will stop tracking if there are no more transactions.
+   */
+  private _shouldStopTracking(): void {
+    if (this._statsByTransaction.size === 0) {
+      this._stopTracking();
+    }
+  }
+
+  /**
+   * Clears all the collected stats
+   */
+  private _reset(): void {
+    this._stallCount = 0;
+    this._totalStallTime = 0;
+    this._lastIntervalMs = 0;
+    this._statsByTransaction.clear();
+  }
+
+  /**
+   * Iteration of the stall tracking interval. Measures how long the timer strayed from its expected time of running, and how
+   * long the stall is for.
+   */
+  private _iteration(): void {
+    const now = timestampInSeconds() * 1000;
+    const totalTimeTaken = now - this._lastIntervalMs;
+
+    if (
+      totalTimeTaken >=
+      LOOP_TIMEOUT_INTERVAL_MS + this._minimumStallThreshold
+    ) {
+      const stallTime = totalTimeTaken - LOOP_TIMEOUT_INTERVAL_MS;
+      this._stallCount += 1;
+      this._totalStallTime += stallTime;
+
+      for (const [transaction, value] of this._statsByTransaction.entries()) {
+        const longestStallTime = Math.max(
+          value.longestStallTime ?? 0,
+          stallTime
+        );
+
+        this._statsByTransaction.set(transaction, {
+          ...value,
+          longestStallTime,
+        });
+      }
+    }
+
+    this._lastIntervalMs = now;
+
+    if (this.isTracking) {
+      this._timeout = setTimeout(
+        this._iteration.bind(this),
+        LOOP_TIMEOUT_INTERVAL_MS
+      );
+    }
+  }
+
+  /**
+   * Deletes leaked transactions (Earliest transactions when we have more than MAX_RUNNING_TRANSACTIONS transactions.)
+   */
+  private _flushLeakedTransactions(): void {
+    if (this._statsByTransaction.size > MAX_RUNNING_TRANSACTIONS) {
+      const leakedTransactions = Array.from(
+        this._statsByTransaction.keys()
+      ).slice(0, this._statsByTransaction.size - MAX_RUNNING_TRANSACTIONS);
```

Comment:
Feel like this having this logic means that we should probably be using an LRU cache instead of a Map.

But I guess it's fine to keep it as it is. An explicit for loop will be more efficient than slice + forEach here.

```ts
let counter = 0;
const transactions = this._statsByTransaction.keys();
const len = this._statsByTransaction.size - MAX_RUNNING_TRANSACTIONS
for (let t of transactions) {
    if (counter > len) break;
    counter++;
    this._statsByTransaction.delete(t);
}
```